### PR TITLE
Add transaction-payment runtime API V3

### PR DESCRIPTION
**Pull request by notlesh**
This PR is a companion for https://github.com/paritytech/substrate/pull/13110. It adds the same API definitions to `pallet-transaction-payment` introduced there. The API version is being rev'ed from V2 -> V3.

I have not been able to get this to work and could use a pointer.

When run my local changes here against a node/runtime supporting the new V3, the entire `api.call.transactionPaymentApi` is missing (`undefined`), likely the same symptom described in the PR above. It seems to work fine for V2.

I *think* I need to run `yarn polkadot-types-from-chain` to generate some code and put it in the right place, but I'm not sure exactly how to invoke this. I could be barking up the wrong tree, of course.

### CODE REVIEW

**Comment by jacogr** [comment]
Weird, at quick glance it does add the right things. (The typegen is only needed for TS developers, so it doesn't need to be run)

How are you testing this against a local node? (Which steps?)

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:61]:
```suggestion
    description: 'Query the output of the current WeightToFee given some input',
```

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:74]:
```suggestion
    description: 'Query the output of the current LengthToFee given some input',
```

**Review by jacogr — COMMENTED** [review]
Some small linting fixes on quotes, aka ' vs "

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:92]:
```suggestion
      methods: objectSpread(
        {},
        V3_QUERY_WEIGHT_TO_FEE,
        V3_QUERY_LENGTH_TO_FEE,
        V2_V3_SHARED_QUERY_INFO,
        V1_V2_V3_SHARED_PAY
      ),
```

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:98]:
```suggestion
      methods: objectSpread(
        {},
        V2_V3_SHARED_QUERY_INFO,
        V1_V2_V3_SHARED_PAY
      ),
```

**Review by jacogr — COMMENTED** [review]
Just spread into empty object (otherwise it would actually clobber the existing)

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:135]:
```suggestion
      methods: objectSpread(
        {},
        V3_QUERY_WEIGHT_TO_FEE,
        V3_QUERY_LENGTH_TO_FEE,
        V2_V3_SHARED_QUERY_INFO,
        V1_V2_V3_SHARED_CALL
      ),
```

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:141]:
```suggestion
      methods: objectSpread(
        {},
        V2_V3_SHARED_QUERY_INFO,
        V1_V2_V3_SHARED_CALL
      ),
```

**Review by jacogr — COMMENTED** [review]
Apply same "to-empty-spread" functionality for TransactionPaymentCallApi

**Comment by jacogr** [comment]
After applying the above locally and running against your node -

image

image

So seems fine.

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:59]: This one is now missing....

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:91]: ... and this one is only for `TransactionPaymentCallApi` (so the naming mismatches)

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:46]: `query_info` is now missing in yours (see the point below)

File: packages/types/src/interfaces/payment/runtime.ts
Comment (jacogr) [inline packages/types/src/interfaces/payment/runtime.ts:92]: ... this contains `query_call_info` which only applies to `TransactionPaymentCallApi` below. (The above call is not missing in this version)

Comment (notlesh) [reply packages/types/src/interfaces/payment/runtime.ts:109]: Good catch, should be fixed in 9ff401e.

**Comment by notlesh** [comment]
Thanks for the help, Jaco! I believe addressed the `query_call_info` issue, PTAL.

However, I haven't been able to set up an override for `@polkadot.js/types` to test this in any way. I've tried a few things but am really just shooting in the dark, is there some documentation on doing this? (I'm a typescript noob, btw).

**Comment by jacogr** [comment]
Depends on how you want to test. So for my tests above, I have this structure - `<ROOT>/api` and `<ROOT>/apps` (i.e. bot repos in the same top-level)

1. Apply the changes in api
2. inside api I do a `yarn polkadot-dev-copy-to apps` (this compiles all and replaces the `node_modules` inside `apps`)
3. Inside apps, I do a `yarn start`

So basically the `polkadot-dev-copy-to` command is useful to get stuff into other projects. (The first `..` is assumed, but any depth can be achieved, for sanity just made it simple to assume the same top-level structure in those scripts)

**Comment by notlesh** [comment]
Thanks! I was able to test the changes, LGTM now!

**Review by jacogr — COMMENTED** [review]
Looks all-ok. Will wait for the Substrate PR to be merged before merging this one (just in-case there are some changes required there)

**Comment by ggwpez** [comment]
Substrate just merged https://github.com/paritytech/substrate/pull/13110

**Review by jacogr — APPROVED** [review]
Thank you.

**Pull request merged**
